### DISCUSSION TITLE
How to stop repeating my Java version?

### DISCUSSION THREAD

**Discussion by Mahoney** [sdkman/sdkman-cli#1262] [Q&A]
I'm currently declaring my Java version in 3 places:
* `.sdkmanrc` as `java=17.0.9-tem`
* `build.gradle.kts` as `jvmToolchain(17)`
* `Dockerfile` as `FROM eclipse-temurin:17.0.9_9-jdk-focal`

That seems bad... anyone got any tricks for avoiding this redeclaration?

I think I can interpolate into the `Dockerfile` using a `.env` file containing `JAVA_VERSION=17.0.9_9`, and I can probably hack something in Gradle to read the value from a different file, but it looks like there's no way to parameterise `.sdkmanrc`?

**Comment by goldenflo** [comment]
For Windows:
Open the Start menu and search for []("**Environment Variables**."https://vytcdc.us/java-online-training/)
Click on "Edit the system environment variables."
In the System Properties window, click "Environment Variables..."
Under "System variables," find and select the "Path" variable, then click "Edit..."
Remove entries related to Java (e.g., C:\Program Files\Java\jdk1.x.x\bin).
Click "OK" to close each window.
For Unix/Linux/Mac:
Open your shell configuration file: ~/.bashrc, ~/.bash_profile, or ~/.zshrc.
Add the following line at the end, replacing path/to/your/java/bin with the actual path to your Java bin directory:
bash
Copy code
export PATH=path/to/your/java/bin:$PATH
Save and close the file.
Run the command source ~/.bashrc (or source ~/.bash_profile or source ~/.zshrc, depending on your configuration) to apply the changes immediately.
After following these steps, the Java version information should no longer be displayed each time you open a new terminal or command prompt window. Adjust the instructions based on your specific operating system and shell.

**Comment by helpermethod** [comment]
> I'm currently declaring my Java version in 3 places:
> 
> * `.sdkmanrc` as `java=17.0.9-tem`
> * `build.gradle.kts` as `jvmToolchain(17)`
> * `Dockerfile` as `FROM eclipse-temurin:17.0.9_9-jdk-focal`
> 
> That seems bad... anyone got any tricks for avoiding this redeclaration?
> 
> I think I can interpolate into the `Dockerfile` using a `.env` file containing `JAVA_VERSION=17.0.9_9`, and I can probably hack something in Gradle to read the value from a different file, but it looks like there's no way to parameterise `.sdkmanrc`?

The `.sdkmanrc` file loosely follows the [Properties File Format](https://docs.oracle.com/cd/E23095_01/Platform.93/ATGProgGuide/html/s0204propertiesfileformat01.html), so you could probably read the `.sdkmanrc` file into a Properties object and use that in your build.gradle.kts.

On the other hand, if you are already using Toolchains, you may not need an `.sdkmanrc` file at all.

**Reply by Mahoney** [reply]
I like to control the JVM used to run gradle in the first place. I only use Toolchains as a shorthand way of setting up target class format. There's probably a better way.

Gradle's a bit of a red herring here, I shouldn't have mentioned it - I can work out how to read `.sdkmanrc` from Gradle pretty easily. It's more how to share the version between the `Dockerfile` and `.sdkmanrc`.

I worked out one way - use SDKMAN! to install the JVM in the Dockerfile instead of using an existing image:
```dockerfile
FROM alpine

RUN apk add --no-cache \
    bash \
    curl \
    zip

RUN curl -s "https://get.sdkman.io" | bash

COPY --link .sdkmanrc .
RUN bash -c "source ~/.sdkman/bin/sdkman-init.sh && sdk env install"
```
I'm not sure what I think about it...

EDIT - spoke too soon, the above doesn't set JAVA_HOME on the default shell, so doesn't work. Don't have time to work out how to fix that yet!

**Comment by goldenflo** [comment]
To stop repeating your Java version, set the JAVA_HOME environment variable and add the bin directory to the PATH variable. This ensures your system uses the specified Java version without needing constant repetition.

Find your Java installation path:https://vytcdc.us/java-online-training/

On Windows: C:\Program Files\Java\jdk1.x.x
On Linux/Mac: /usr/lib/jvm/jdk1.x.x
Set the JAVA_HOME environment variable:

On Windows:
bash
Copy code
setx JAVA_HOME "C:\Program Files\Java\jdk1.x.x" /M
On Linux/Mac:
bash
Copy code
export JAVA_HOME=/usr/lib/jvm/jdk1.x.x
Add the bin directory to the PATH variable:

On Windows:
bash
Copy code
setx PATH "%PATH%;%JAVA_HOME%\bin" /M
On Linux/Mac:
bash
Copy code
export PATH=$PATH:$JAVA_HOME/bin
Verify the changes:

Open a new terminal/cmd window and type:
bash
Copy code
java -version
Now, you won't need to repeat the Java version setup each time you open a new terminal/cmd window.
